tools/convert_to_hf.py

```diff
@@ -0,0 +1,271 @@
+    """
+    key = key.replace('_', '-')
+    try:
+        return loaded_config[key]
+    except KeyError:
+        key = key.replace('-', '_')
+        try:
+            return loaded_config[key]
+        except KeyError:
+            return default 
+
+def create_config(neox_config):
+    """ take in a loaded yaml from NeoX and assign relevant values to HF config.
+    Returns: GPTNeoXConfig() object
+    """
+    class TokenizerArgs:
+        # kinda hacky. 
+        # this is to get something with the same interface as is used in build_tokenizer()
+        # without diving into loading a neox_args object or using argparse etc.
+        def __init__(self, neox_config):
+            self.make_vocab_size_divisible_by = get_key(neox_config, 'make-vocab-size-divisible-by', default=128)
+            self.model_parallel_size = get_key(neox_config, 'model-parallel-size')
+            self.vocab_file = get_key(neox_config, 'vocab-file')
+            self.merge_file = get_key(neox_config, 'merge-file')
+            self.tokenizer_type = get_key(neox_config, 'tokenizer-type')
+
+            self.rank = 0
+
+
+    args = TokenizerArgs(neox_config)
+    tokenizer = build_tokenizer(args)
+    try: # GPT2TokenizerFast raises NotImplementedError
+        pad_token = tokenizer.pad
+    except:
+        pad_token = 1 # pad defaulting to 1. follows convention from GPT-NeoX-20b tokenizer
+
+    # set all config values.
+    hf_config = GPTNeoXConfig(
+        vocab_size=args.padded_vocab_size,
+        hidden_size=get_key(neox_config, 'hidden-size'),
+        num_hidden_layers=get_key(neox_config, 'num-layers'),
+        num_attention_heads=get_key(neox_config, 'num-attention-heads'),
+        intermediate_size=(get_key(neox_config, 'hidden-size') * 4),
+        hidden_act=get_key(neox_config, 'activation', default='gelu'),
+        rotary_pct=get_key(neox_config, 'rotary-pct', default=1.0),
+        rotary_emb_base=get_key(neox_config, 'rotary-emb-base', default=10000),
+        max_position_embeddings=get_key(neox_config, 'max-position-embeddings'),
+        initializer_range=get_key(neox_config, 'init-method-std', 0.02),
+        layer_norm_eps=get_key(neox_config, 'layernorm-epsilon', 1e-5),
+        use_cache=True,
+        bos_token_id=tokenizer.eod,
+        eos_token_id=tokenizer.eod,
+        tie_word_embeddings=(not get_key(neox_config, 'no-weight-tying', False)),
+        use_parallel_residual=get_key(neox_config, 'gpt-j-residual', False)
+    )
+    return hf_config
+
+
+def convert(input_checkpoint_path, loaded_config, output_checkpoint_path):
+    """ convert a NeoX checkpoint to a HF model format. 
+    should perform model-parallel merging correctly
+    but only supports features allowed by HF GPT-NeoX implementation (e.g. rotary embeddings)
+    """
+
+    hf_config = GPTNeoXConfig()
+
+    # TODO: how to deal with vocab size? need to pass tokenizer and calc on the fly. 
```

Comment:
Has this TODO been resolved? If so, let's remove this. If not, I'm not sure what "need to pass tokenizer and calc on the fly" means

Reply:
Yes sorry, removed it!

This refers to calculating the vocab size with padding to a multiple of 128, and it’s done in this script now.